pkg/virt-launcher/virtwrap/api/converter.go

```diff
@@ -102,6 +103,37 @@ func Convert_v1_Disk_To_api_Disk(diskDevice *v1.Disk, disk *Disk, devicePerBus m
 	return nil
 }
 
+func setDriverCacheMode(disk *Disk, mode v1.DriverCache) error {
+	supportDirectIO := true
+
+	if mode == "" || mode == v1.CacheNone {
+		f, err := os.OpenFile(disk.Source.File, syscall.O_RDONLY|syscall.O_DIRECT, 0)
+		f.Close()
+		if err != nil && !os.IsNotExist(err) {
+			log.Log.Infof("%s file system does not support direct I/O", disk.Source.File)
+			supportDirectIO = false
+		}
+	}
+
+	// if user set a cache mode = 'none' and fs does not support direct I/O then return an error
+	if mode == v1.CacheNone && !supportDirectIO {
+		return fmt.Errorf("Unable to use '%s' cache mode, file system where %s is stored does not support direct I/O", mode, disk.Source.File)
+	}
+
+	// if user did not set a cache mode and fs supports direct I/O then set cache = 'none'
```

Comment:
I'm a little uncertain about what we should be doing (if anything) if cache mode isn't set.

If no option was set, i'd expect the default behavior libvirt/qemu has to be used. We're doing our own defaulting here. I don't know enough about how this feature is used in practice to say whether or not that's an issue. Just something worth pointing out. 

@vladikr any thoughts on this?

Reply:
iirc, libvirt's default is `none` as well,  but it's better for us to set `none` if the device supports it. `none` should be the most consistent for us when considering migrations. This way we don't have to worry about a redundant cache in the kernel (guest and host)